### PR TITLE
construct mlperf qBERT evaluation

### PR DESCRIPTION
**Pull request by deeplearningfromscratch**
A100 에서 [qBERT 재현 instruction](https://www.notion.so/furiosa/bert-v1-1-guide-003a2d827b884e5b84ed2200bb1d6f32) 에 따라 환경을 구성했으며 w/ calibration, w/o calibration 모두 아래와 같은 결과를 확인했습니다.

|    |        our result        | reference accuracy |
|:--:|:------------------------:|:---------------:|
| F1 | 91.05864237161231 (100.20%) | [90.874](https://github.com/mlcommons/inference/blob/e39003a9c4c89a2215db0ca57ad7a57b16f9a785/tools/submission/submission_checker.py#L1119C31-L1119C37)

```
/root/workspace/inference/language/bert/accuracy-squad.py:27: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
{"exact_match": 83.85052034058657, "f1": 91.05864237161231}
Reading examples...
Loading cached features from 'eval_features.pickle'...
Loading LoadGen logs...
Post-processing predictions...
Writing predictions to: /root/workspace/inference/logs/qbert/Offline/20240328_110426UTC/predictions.json
Evaluating predictions...

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
 DEBIAN_FRONTEND=noninteractive apt-get update && apt-get install libgl1 libglib2.0-0 -y
 ```
 
+
+## Submission Model
+### Wi8Ai8KVi8 quantized BERT(qBERT)
```
Comment (dkim-furiosa) [inline README.md]:
```suggestion

### Wi8Ai8KVi8 quantized BERT (qBERT)
```

* https://english.stackexchange.com/questions/5987/is-there-any-rule-for-the-placement-of-space-after-and-before-parentheses
* https://www.markdownguide.org/basic-syntax/#heading-best-practices

File: README.md [hunk 2]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+
+- Evaluation result
+
+    |    |        our result        | reference accuracy |
+    |:--:|:------------------------:|:---------------:|
```
Comment (dkim-furiosa) [inline README.md]:
```suggestion
    |    |         our result          | reference accuracy |
    |:--:|:---------------------------:|:------------------:|
```

File: README.md [hunk 3]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+    | F1 | 91.05864237161231 (100.20%) | [90.874](https://github.com/mlcommons/inference/blob/e39003a9c4c89a2215db0ca57ad7a57b16f9a785/tools/submission/submission_checker.py#L1119C31-L1119C37)
+
+- To reprocude the result, you can run one of the following commands:
```
Comment (dkim-furiosa) [inline README.md]:
```suggestion
- To reproduce the result, you can run one of the following commands:
```

File: README.md [hunk 4]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+        . scripts/build_qbert_env.sh
+        . scripts/eval_qbert.sh
+        ```
+
+    3. 
+        ```
+        export SCENARIO=Offline # SCENARIO is one of [Offline, SingleStream, MultiStream, Server]
+        export N_COUNT=10833   # N_COUNT is a number between [1, 10833]
+        export CALIBRATE=false # CALIBRATE is one of [false, true]
+        export N_CALIB=100 # N_CALIB is a number between [1, 100]
```
Comment (dkim-furiosa) [inline README.md]: 1번과 3번 둘 다 동일한 `make qbert`를 실행하는데, 왜 3번에서만 환경 변수 설정이 있는지요?

File: README.md [hunk 5]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+        make qbert
+        ```
+
+### Wi8Ai8KVi8 quantized GPT-J(qGPT-J)
```
Comment (dkim-furiosa) [inline README.md]:
```suggestion
### Wi8Ai8KVi8 quantized GPT-J (qGPT-J)
```

File: README.md [hunk 6]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+
+TBA
+
+### Wi8Ai8KVi8 quantized LLaMA2-70b(qLLaMA2-70b)
```
Comment (dkim-furiosa) [inline README.md]:
```suggestion
### Wi8Ai8KVi8 quantized LLaMA2-70b (qLLaMA2-70b)
```

File: README.md [hunk 7]
```diff
@@ -16,8 +16,49 @@ apt-get update && apt-get install build-essential -y
...
+
+TBA
+
+
 ## How to run end-to-end evaluation
-Ene-to-end(E2E) evaluation is the process of downloading models and dataset, building a Python environment, and performing model accuracy evaluation. E2E scripts are developed based on [f9a643c](https://github.com/mlcommons/inference/commit/f9a643c0a0e920588da1b51a1d822e1071a9dbec). 
+End-to-end(E2E) evaluation is the process of downloading models and dataset, building a Python environment, and performing model accuracy evaluation. E2E scripts are developed based on [f9a643c](https://github.com/mlcommons/inference/commit/f9a643c0a0e920588da1b51a1d822e1071a9dbec).
```
Comment (dkim-furiosa) [inline README.md:54]:
```suggestion

End-to-end(E2E) evaluation is the process of downloading models and dataset, building a Python environment, and performing model accuracy evaluation. E2E scripts are developed based on [f9a643c](https://github.com/mlcommons/inference/commit/f9a643c0a0e920588da1b51a1d822e1071a9dbec). 
```

File: language/bert/quantization/calibrate.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
+import argparse
+import json
+import os
+import pickle
+import sys
+
+sys.path.insert(0, os.getcwd())
```
Comment (dkim-furiosa) [inline language/bert/quantization/calibrate.py]: 이 줄이 필요한가요? 그리고, 실행할 때 현재 디렉토리가 어디인가에 따라 효과가 달라져서 불안정할 듯합니다.

File: language/bert/quantization/__init__.py [hunk 9]
```diff
@@ -1 +1,33 @@
-from .get_quant_model import *
+import os
+
+import torch
+import yaml
+from quantization.custom_symbolic_trace import custom_symbolic_trace
```
Comment (dkim-furiosa) [inline language/bert/quantization/__init__.py]: `quantization` 로컬 패키지가 그 위 패키지와 묶이는 건 어색한 듯합니다. 그리고, 나머지 파일들과 순서에 일관성이 없습니다.

File: language/bert/quantization/calibrate.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        hidden_dropout_prob=config_json["hidden_dropout_prob"],
+        hidden_size=config_json["hidden_size"],
+        initializer_range=config_json["initializer_range"],
+        intermediate_size=config_json["intermediate_size"],
+        max_position_embeddings=config_json["max_position_embeddings"],
+        num_attention_heads=config_json["num_attention_heads"],
+        num_hidden_layers=config_json["num_hidden_layers"],
+        type_vocab_size=config_json["type_vocab_size"],
+        vocab_size=config_json["vocab_size"],
+    )
```
Comment (dkim-furiosa) [inline language/bert/quantization/calibrate.py]: `config_json`이 사전이라면, 다음이 동작하는지 확인해서 적용 부탁드립니다.

```suggestion
    config = BertConfig(**config_json)
```

여기 뿐만 아니라 설정을 읽어들여서 설정을 만드는 다른 부분들도 비슷한 방식으로 지금보다 간명하게 기술할 수 있을지 검토 바랍니다.

File: language/bert/quantization/calibrate.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    for feature in cal_features:
+        data_list.append(
+            {
+                "input_ids": torch.LongTensor(feature.input_ids),
+                "attention_mask": torch.LongTensor(feature.input_mask),
+                "token_type_ids": torch.LongTensor(feature.segment_ids),
+            }
+        )
+
+    return DataLoader(data_list[:n_calib], batch_size=batch_size)
```
Comment (dkim-furiosa) [inline language/bert/quantization/calibrate.py]:
```suggestion
    data_list = [
        {
            "input_ids": torch.LongTensor(feature.input_ids),
            "attention_mask": torch.LongTensor(feature.input_mask),
            "token_type_ids": torch.LongTensor(feature.segment_ids),
        }
        for feature in cal_features[:n_calib]
    ]

    return DataLoader(data_list, batch_size=batch_size)
```

File: README.md (hunk 4 above)
Comment (deeplearningfromscratch) [reply README.md]: 이 부분은 필요한 경우 변경할 수 있는 환경 변수들을 설명하기 위해 넣었습니다.

Comment (dkim-furiosa) [reply README.md]: 해당 목록에 속한 항목들은 첫 문장 "To reprocude the result, you can run one of the following commands"과 밀접하게 연관되어야 할 텐데, 설정 변경은 이 목록에 어울리지 않는 듯합니다. 설정 변경에 대해 안내하는 별도 문단을 만들고 이 항목도 그 쪽으로 옮기는 게 매끄럽지 않을까 싶습니다.

**Comment by deeplearningfromscratch** [comment]
(조금이라도) 모델 동작/정확도에 영향을 줄 것으로 예상되는 변경 commit 에 대해서 validation dataset 및 calibration 각 **subset 100** 개에 대해서 accuracy 를 확인하고자 합니다. 
현재 사용할 수 있는 A6000 GPU 에서는 동 accuracy 가 `{"exact_match": 93.0, "f1": 94.86666666666667}` 을 만족하는지 체크합니다.

File: language/bert/quantization/calibrate.py (hunk 8 above)
Comment (deeplearningfromscratch) [reply language/bert/quantization/calibrate.py]: `quantization` 안에서 내부 모듈을 불러오기 위해서 필요합니다. 절대 경로로 표시하면 어떨까요?

File: README.md (hunk 1 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다.

File: README.md (hunk 2 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다. 추가로 reference accuracy 를 accuracy target 으로 바꿨습니다.

File: README.md (hunk 3 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다.

File: README.md (hunk 5 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다.

File: README.md (hunk 6 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다.

File: README.md (hunk 7 above)
Comment (deeplearningfromscratch) [reply README.md:54]: https://github.com/furiosa-ai/inference/pull/12/commits/4fbd511402a7c47bef5403c76bbd031b088319be 반영했습니다.

File: language/bert/quantization/__init__.py (hunk 9 above)
Comment (deeplearningfromscratch) [reply language/bert/quantization/__init__.py]: https://github.com/furiosa-ai/inference/pull/12/commits/dc6036a0552efb0d2cf6f78d9288234adcbae6d6 반영했습니다.

File: language/bert/quantization/calibrate.py (hunk 11 above)
Comment (deeplearningfromscratch) [reply language/bert/quantization/calibrate.py]: https://github.com/furiosa-ai/inference/pull/12/commits/a9b6128656620f22fec2e635960b67191f119bf6 반영했습니다.

File: language/bert/quantization/calibrate.py (hunk 10 above)
Comment (deeplearningfromscratch) [reply language/bert/quantization/calibrate.py]: https://github.com/furiosa-ai/inference/pull/12/commits/a9b6128656620f22fec2e635960b67191f119bf6 반영했습니다. 
https://github.com/furiosa-ai/inference/pull/12/commits/6ff06fb23e12eae75ffacb1b099dd439e33d0b73 반영했습니다.

File: README.md (hunk 4 above)
Comment (deeplearningfromscratch) [reply README.md]: https://github.com/furiosa-ai/inference/pull/12/commits/4a76712f46267369e8886823def95e2b1627ab0d 반영했습니다.

**Review by hyunsik — APPROVED** [review]
네 저는 전체적으로 좋은 것 같습니다. 몇 가지 코멘트 드렸는데요. 반영 후 머지하셔도 좋을 것 같습니다.

File: Makefile [hunk 12]
```diff
@@ -1,3 +1,9 @@
+# end-to-end quantized model evaluation
+.PHONY: qbert
+qbert:
+	-bash scripts/build_qbert_env.sh
+	-bash scripts/eval_qbert.sh
```
Comment (hyunsik) [inline Makefile]: 환경 설정은 1회성이라 eval_qbert.sh 만 남겨두고 build_qbert_env.sh 는 제거하는 것이 좋을 것 같습니다.

```suggestion
.PHONY: qbert
qbert:
	-bash scripts/eval_qbert.sh
```

File: README.md [hunk 13]
```diff
@@ -16,8 +16,52 @@ apt-get update && apt-get install build-essential -y
...
+        ```
+        
+- Some parameters are configurable like below.
+
+    ```
+        export SCENARIO=Offline # SCENARIO is one of [Offline, SingleStream, MultiStream, Server]
+        export N_COUNT=10833   # N_COUNT is a number between [1, 10833]
+        export CALIBRATE=false # CALIBRATE is one of [false, true]
+        export N_CALIB=100 # N_CALIB is a number between [1, 100]
+        make qbert
```
Comment (hyunsik) [inline README.md]: 아래 부분은 수정이 필요할 것 같습니다. "to reproduce the result" 라는 내용이 포함되어 있으면 위 결과를 동일하게 재현할 수 있는 각종 상세가 주어져야 할 것 같습니다. 기본설정으로 주어져있다 하더라도 명시적으로 설정하는 방식의 커맨드를 아래 써주시면 좋을 것 같습니다. 아래 제가 넣은 설명 한번 보시고 검토 부탁드립니다. 잘 모르는 내용은 ?? 로 표시해두었습니다.

````suggestion
To reproduce the above accuracy result, please run as follow:

```
. scripts/build_qbert_env.sh  # You can skip this step if you already setup the environment.
make qbert
```

For other evaluations with different settigs, you can also change the following environment variables. Each configuration is defined as follows:
* `SCENARIO` is a MLPerf scenario to evaluate the accuracy. It can be one of Offline, SingleStream, MultiStream, or Server.
* `N_COUNT`: the number of data to evaluate the accuracy ([1, 10833])
* `CALIBRATE`: the calibration is involved in the evaluation step if true. (default: ???)
* `N_CALIB`: the number of data to calibrate the quantized model ([1, 100])
````

File: Makefile (hunk 12 above)
Comment (deeplearningfromscratch) [reply Makefile]: 비슷한 코멘트여서 https://github.com/furiosa-ai/inference/pull/16#discussion_r1562085496 에서 아래와 같이 의견남겼습니다. 의견 부탁드리겠습니다.

> 제가 의도했던 것은 매 모델 평가 시 매번 환경 구성을 진행하는 것이 기본이고, 만약 환경 구성을 한 번하고 실험을 반복적으로 하고자하면 오히려 bash scripts/eval_qgpt-j.sh 를 반복해서 실행하자는 것이었습니다.
> 그렇게 생각한 이유는 한 번 conda env. activate 된 상황에서 다른 어떠한 이유로 환경이 바뀔 수가 있고, 사용자가 인지하지 못한 이유로 평가 결과가 바뀌는 것을 막기 위함이었습니다.
>

File: language/bert/quantization/calibrate.py (hunk 8 above)
Comment (dkim-furiosa) [reply language/bert/quantization/calibrate.py]: relative import를 사용해서 해결할 수 있는지 확인 부탁드립니다: https://realpython.com/absolute-vs-relative-python-imports/#relative-imports

File: language/bert/quantization/calibrate.py (hunk 10 above)
Comment (dkim-furiosa) [reply language/bert/quantization/calibrate.py]: 다음은 추측에 기반해서 작성하였습니다. 정래 님이 생각하신 `get_kwargs` 역할에 맞는지 확인하고, 맞으면 적용 부탁드립니다:

```python
import inspect
from typing import Any, Mapping


def get_kwargs(fn, config_dict: Mapping[str, Any]):
    params = inspect.signature(fn).parameters
    return {
        k: v
        for k, v in conifg_dict
        if k in params and params[k].annotation in (inspect.Parameter.emtpy, type(v))
    }
```

Comment (deeplearningfromscratch) [reply language/bert/quantization/calibrate.py]: [review: simpler is better](https://github.com/furiosa-ai/inference/pull/12/commits/a0e352be5952ff6f2e7711768be88a91bc3601db) 반영했습니다. 훨씬 깔끔해졌네요.

File: language/bert/quantization/calibrate.py (hunk 8 above)
Comment (deeplearningfromscratch) [reply language/bert/quantization/calibrate.py]: [review: relative import](https://github.com/furiosa-ai/inference/pull/12/commits/134653452a9384cd2dcf6d9b06d68ab6dad35ba6) 반영했습니다. 너무 어렵게 생각했던듯 합니다.

File: Makefile (hunk 12 above)
Comment (deeplearningfromscratch) [reply Makefile]: [review: make instruction simple](https://github.com/furiosa-ai/inference/pull/12/commits/c5fe20f6a85c8641502fec12cd3b41ee03fda9fc) 반영했습니다. 다시 한 번 생각해보니까 말씀해주신대로 구성하면 훨씬 깔끔한 것 같습니다.

File: README.md (hunk 13 above)
Comment (deeplearningfromscratch) [reply README.md]: [review: make instruction simple](https://github.com/furiosa-ai/inference/pull/12/commits/c5fe20f6a85c8641502fec12cd3b41ee03fda9fc) 반영했습니다.

**Review by dkim-furiosa — APPROVED** [review]
저는 이제 좋은 듯합니다. 수고하셨습니다.

**Comment by dkim-furiosa** [comment]
PR #12 머지되면 그 PR 리뷰에 따른 변경 사항을 이 PR에도 적용 부탁드립니다.
